java/ql/src/Security/CWE/CWE-378/TempDirHijackingVulnerability.ql

```diff
@@ -0,0 +1,427 @@
+  result =
+    any(MethodAccess existanceCheck |
+      (
+        existanceCheck.getMethod() instanceof MethodFileExists
+        or
+        existanceCheck.getMethod() instanceof MethodFileIsDirectory
+      )
+    ) and
+    result.directlyControls(s.getEnclosingStmt(), false)
+}
+
+/**
+ * A guard that is verifying that the directory is sucsessfully created, throwing when it is not created.
+ */
+private Guard directoryCreationGuard(ThrowStmt s) {
+  result = any(MethodAccess creationCheck |
+    creationCheck.getMethod() instanceof MethodFileCreatesDirs
+  ) and
+  result.directlyControls(s.getEnclosingStmt(), false)
+}
+
+/**
+ * A guard that is safely verfying that a directory is created.
+ * 
+ * 'Safe' means that the directory is guarnteed to have been created, and the directory did not already exist.
+ */
+private Guard safeDirectoryCreationGuard(ThrowStmt s) {
+  result = directoryCreationGuard(s) and
+  // This guard is not 'safe' if a `directoryExistsGuard` is also guarding this throw statement
+  not exists(directoryExistsGuard(s))
+}
+
+/**
+ * An expression that will create a directory without throwing an exception if a file/directory already exists.
+ */
+private predicate isNonThrowingDirectoryCreationExpression(
+  Expr fileInstanceExpr, MethodAccess creationCall
+) {
+  creationCall.getMethod() instanceof MethodFileCreatesDirs and
+  creationCall.getQualifier() = fileInstanceExpr and
+  // Check for 'throwing creations' and ensure that this call is not used in a throwing case.
+  if creationCall.(Guard).directlyControls(_, _) then // The creation call is used in a guard
+    not creationCall = safeDirectoryCreationGuard(_) // Ensure that the guard is insufficient.
+  else any() // The creation call is not used in a guard. Thus is not a 'throwing creation'.
+  or
+  // Recursively look for methods that encapsulate the above.
+  // Thus, the use of 'helper directory creation methods' are still considered
+  // when assessing if an `unsafeUse` is present or not.
+  exists(Method m, int argumentIndex, Expr arg |
+    arg.(VarAccess).getVariable() = m.getParameter(argumentIndex) and
+    isNonThrowingDirectoryCreationExpression(any(Expr e2 | DataFlow::localExprFlow(arg, e2)),
+      creationCall)
+  |
+    m.getAReference().getArgument(argumentIndex) = fileInstanceExpr
+  )
+}
+
+private class MethodFileDelete extends Method {
+  MethodFileDelete() {
+    this.getDeclaringType() instanceof TypeFile and
+    this.hasName("delete")
+  }
+}
+
+private class MethodFileIsDirectory extends Method {
+  MethodFileIsDirectory() {
+    this.getDeclaringType() instanceof TypeFile and
+    this.hasName("isDirectory")
+  }
+}
+
+private class MethodFileExists extends Method {
+  MethodFileExists() {
+    this.getDeclaringType() instanceof TypeFile and
+    this.hasName("exists")
+  }
+}
+
+predicate isDeleteFileExpr(Expr expr) {
+  expr = any(MethodFileDelete m).getAReference().getQualifier()
+}
```

Comment:
I think I may want to expand this to include delete wrapper functions. For example something like the following:

```java
void deleteSafely(File file) throws IOException {
    if (!file.delete()) throw new IOException("Failed to delete file");
}
```